nf_core/modules.py

```diff
@@ -231,3 +237,481 @@ def has_valid_pipeline(self):
+                nfcore_modules = [nfcore_modules[idx]]
+            except ValueError as e:
+                log.error("Could not find the given module!")
+                sys.exit(1)
+
+        # Lint local modules
+        self.lint_local_modules(local_modules)
+
+        # Lint nf-core modules
+        self.lint_nfcore_modules(nfcore_modules)
+
+        if print_results:
+            self._print_results(show_passed=show_passed)
+
+        return {"passed": self.passed, "warned": self.warned, "failed": self.failed}
+
+    def lint_local_modules(self, local_modules):
+        """
+        Lint a local module
+        Only issues warnings instead of failures
+        """
+        for mod in local_modules:
+            self.lint_main_nf(mod)
+
+    def lint_nfcore_modules(self, nfcore_modules):
+        """
+        Lint nf-core modules
+        For each nf-core module, checks for existence of the files
+        - main.nf
+        - meta.yml
+        - functions.nf
+        And verifies that their content.
+
+        If the linting is run for modules in the central nf-core/modules repo
+        (repo_type==modules), files that are relevant for module testing are
+        also examined
+        """
+        # Iterate over modules and run all checks on them
+        for mod in nfcore_modules:
+            if "TOOL/SUBTOOL" in mod:
+                continue
+            module_name = mod.split(os.sep)[-1]
+
+            # Lint the main.nf file
+            inputs, outputs = self.lint_main_nf(os.path.join(mod, "main.nf"))
+
+            # Lint the meta.yml file
+            self.lint_meta_yml(os.path.join(mod, "meta.yml"), module_name, inputs=inputs, outputs=outputs)
+
+            # Lint the functions.nf file
+            self.lint_functions_nf(os.path.join(mod, "functions.nf"))
+
+            if self.repo_type == "modules":
+                self.lint_module_tests(mod, module_name)
+
+    def lint_module_tests(self, mod, module_name):
+        """ Lint module tests """
+        # Extract the software name
+        software = mod.split("software" + os.sep)[1]
+
+        # Check if test directory exists
+        test_dir = os.path.join(self.dir, "tests", "software", software)
+        if os.path.exists(test_dir):
+            self.passed.append("Test directory exsists for {}".format(software))
+        else:
+            self.failed.append("Test directory is missing for {}: {}".format(software, test_dir))
+            return
+
+        # Lint the test main.nf file
+        test_main_nf = os.path.join(test_dir, "main.nf")
+        if os.path.exists(test_main_nf):
+            self.passed.append("test main.nf exists for {}".format(software))
+        else:
+            self.failed.append("test main.nf doesn't exist for {}".format(software))
+
+        # Lint the test.yml file
+        test_yml_file = os.path.join(test_dir, "test.yml")
+        try:
+            with open(test_yml_file, "r") as fh:
+                test_yml = yaml.safe_load(fh)
```

Comment:
Maybe we should have a JSON Schema that describes the expected format of these files... 🤔  Then we could properly validate them fairly easily..

Reply:
Yes that might be a good options .... although I'm wondering whether it's worth it, considering that we can mainly check for the keys `name`, `command`, `tags` and `files`. Everything else is (can be) pretty different from test to test.

Reply:
👌🏻 - my brain is too full of JSON schema these days 😆